File: 3614. Process String with Special Operations II.py
class Solution:
    def processStr(
        self,
        input_string: str,
        target_index: int
    ) -> str:

        string_length = len(input_string)

        # Stores the processed string length
        # after each operation
        processed_lengths = [0] * string_length

        current_length = 0

        # Forward pass
        for current_index in range(string_length):

            current_character = (
                input_string[current_index]
            )

            if 'a' <= current_character <= 'z':

                current_length += 1

            elif current_character == '*':

                if current_length > 0:
                    current_length -= 1

            elif current_character == '#':

                current_length *= 2

            elif current_character == '%':

                pass

            processed_lengths[current_index] = (
                current_length
            )

        if target_index >= current_length:
            return '.'

        # Backward pass
        for current_index in range(
            string_length - 1,
            -1,
            -1
        ):

            current_character = (
                input_string[current_index]
            )

            length_before_operation = (
                0 if current_index == 0
                else processed_lengths[
                    current_index - 1
                ]
            )

            if 'a' <= current_character <= 'z':

                if (
                    target_index ==
                    length_before_operation
                ):
                    return current_character

            elif current_character == '#':

                if length_before_operation > 0:

                    target_index %= (
                        length_before_operation
                    )

            elif current_character == '%':

                target_index = (
                    length_before_operation
                    - 1
                    - target_index
                )

        return '.'

File: 3614. Process String with Special Operations II.cpp
class Solution {
public:

    char processStr(
        string inputString,
        long long targetIndex
    ) {

        int stringLength =
            inputString.size();

        // Stores the length of the
        // processed string after each operation
        vector<long long> processedLengths(
            stringLength
        );

        long long currentLength = 0;

        // Forward pass:
        // Compute the length after each operation
        for (int currentIndex = 0;
             currentIndex < stringLength;
             currentIndex++) {

            char currentCharacter =
                inputString[currentIndex];

            if (currentCharacter >= 'a' &&
                currentCharacter <= 'z') {

                currentLength++;
            }
            else if (currentCharacter == '*') {

                if (currentLength > 0) {
                    currentLength--;
                }
            }
            else if (currentCharacter == '#') {

                currentLength *= 2;
            }
            else if (currentCharacter == '%') {

                // Length does not change
            }

            processedLengths[currentIndex] =
                currentLength;
        }

        // If k is outside the final string
        if (targetIndex >= currentLength) {
            return '.';
        }

        // Backward pass:
        // Trace the character at index k
        for (int currentIndex =
                 stringLength - 1;
             currentIndex >= 0;
             currentIndex--) {

            char currentCharacter =
                inputString[currentIndex];

            long long lengthAfterOperation =
                processedLengths[currentIndex];

            long long lengthBeforeOperation =
                (currentIndex == 0)
                ? 0
                : processedLengths[
                    currentIndex - 1
                  ];

            // Letter insertion
            if (currentCharacter >= 'a' &&
                currentCharacter <= 'z') {

                if (targetIndex ==
                    lengthBeforeOperation) {

                    return currentCharacter;
                }
            }

            // String duplication
            else if (
                currentCharacter == '#'
            ) {

                if (lengthBeforeOperation > 0) {

                    targetIndex %=
                        lengthBeforeOperation;
                }
            }

            // String reversal
            else if (
                currentCharacter == '%'
            ) {

                targetIndex =
                    lengthBeforeOperation -
                    1 -
                    targetIndex;
            }

            // '*' does not need
            // special handling here
        }

        return '.';
    }
};
